File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] charCount = new int[26];
        int maxCharCount = 0;
        int start = 0;
        int longest = 0;

        for (int end = 0; end < s.length(); end++) {
            charCount[s.charAt(end) - 'A']++;
            maxCharCount = Math.max(maxCharCount, charCount[s.charAt(end) - 'A']);

            

            // Explicitly check if the replacements needed exceed k
            if (end - start + 1 - maxCharCount > k) {
                charCount[s.charAt(start) - 'A']--;
                start++;
            }

            longest = Math.max(longest, end - start + 1);
        }

        return longest;
    }
}

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        freq = {}
        max_freq = 0
        res = 0

        l = 0
        for r, c in enumerate(s):
            freq[c] = freq.get(c, 0) + 1

            while (r - l + 1) - max(freq.values()) > k:
                freq[s[l]] -= 1
                l += 1
            res = max(res, (r - l + 1))

        return res